### PR TITLE
Accommodate altered semantics of `cudf::lists::contains()`

### PR DESCRIPTION
**Pull request by mythrocks**
Depends on rapidsai/cudf/pull/9510.
Depends on rapidsai/cudf/pull/9901.

`spark-rapids` relies on [`libcudf`'s](https://github.com/rapidsai/cudf/) `cudf::lists::contains()` for its implementation of `array_contains()`.

rapidsai/cudf/pull/9510 changes the semantics of `lists::contains()`, with regard to rows containing nulls. Specifically, if a list row contains at least one null element, and is found not to contain the search key, `libcudf` will now return `false` instead of `null`.
SparkSQL expects to return `null` in those cases.

This commit accommodates the change in `libcudf's` semantics, to keep its own existing behaviour.

### CODE REVIEW

**Comment by mythrocks** [comment]
build

**Comment by mythrocks** [comment]
This PR is ready for review, although it's still in draft. It can't be merged until the dependencies in CUDF have been merged.

**Comment by mythrocks** [comment]
(Modified for changed CUDF function name. `copyWithValidity` -> `copyWithBooleanColumnAsValidity`.)

**Comment by mythrocks** [comment]
Re-tested after changes to dependencies in `libcudf`. Will rebuild after rapidsai/cudf/pull/9510 is merged.

File: integration_tests/src/main/python/array_test.py [hunk 1]
```diff
@@ -99,10 +99,15 @@ def test_orderby_array_of_structs(data_gen):
 def test_array_contains(data_gen):
     arr_gen = ArrayGen(data_gen)
     lit = gen_scalar(data_gen, force_no_nulls=True)
-    assert_gpu_and_cpu_are_equal_collect(lambda spark: two_col_df(
-        spark, arr_gen, data_gen).select(array_contains(col('a'), lit.cast(data_gen.data_type)),
-                                         array_contains(col('a'), col('b')),
-                                         array_contains(col('a'), col('a')[5])), no_nans_conf)
+
+    def get_input(spark):
```
Comment (revans2) [inline integration_tests/src/main/python/array_test.py:103]: nit: I always feel much better about a fixes like this if the tests weren't changed. They look to be identical to before, but with a lot of formatting changes.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala [hunk 2]
```diff
@@ -223,17 +223,43 @@ case class GpuArrayContains(left: Expression, right: Expression)
...
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuScalar): ColumnVector = {
+    val contains = lhs.getBase.listContains(rhs.getBase)
+    withResource(contains) { containsCV =>
+      val containsNull = lhs.getBase.listContainsNulls()
+      withResource(containsNull) { containsNullCV =>
+        val notContainsNull = containsNullCV.not
+        withResource(notContainsNull) { notContainsNullCV =>
+          val validity = containsCV.or(notContainsNullCV)
+          withResource(validity) { validityCV =>
+            containsCV.copyWithBooleanColumnAsValidity(validityCV)
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala:235]: nit: we could save some memory by closing some of these sooner. Also once `listContains` returns everything after that is common. We should probably turn that into a separate method.

**Review by revans2 — COMMENTED** [review]
Just a few minor nits

File: integration_tests/src/main/python/array_test.py (hunk 1 above)
Comment (mythrocks) [reply integration_tests/src/main/python/array_test.py:103]: Thanks for reviewing, @revans2.
I'd be willing to undo this change, if we could fix the original formatting. I found the previous formatting very hard to parse.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala (hunk 2 above)
Comment (mythrocks) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala:235]: Oh, that's an interesting thought. I'll take a crack at it.

File: integration_tests/src/main/python/array_test.py (hunk 1 above)
Comment (revans2) [reply integration_tests/src/main/python/array_test.py:103]: Like I said it is a nit. If you have strong feelings just keep it.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala (hunk 2 above)
Comment (mythrocks) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala]: Sorry for the delay. Needed to get my head around `withResource()`.
Now, intermediate results are closed early. A new helper function has been added.

**Comment by mythrocks** [comment]
rapidsai/cudf/pull/9510 was just committed. I've pulled this PR out of draft.

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
Even though rapidsai/cudf/pull/9510 has not gone all the way through CI yet. I think we can still test/commit this. Because the extra processing should be a noop if listContains is behaving the old way too.

**Comment by revans2** [comment]
oops I was wrong. It needs a new API that was missing before.

**Comment by tgravescs** [comment]
build

**Comment by mythrocks** [comment]
Merging this to head off impending CI breakage.

**Comment by mythrocks** [comment]
Thank you for the review, @revans2. @tgravescs, thank you for confirming that the CI has passed with the new `libcudf` artifact.
